test/mpi/buffer.cc

```diff
@@ -0,0 +1,99 @@
+/**
+ * Copyright (C) Mellanox Technologies Ltd. 2021.  ALL RIGHTS RESERVED.
+ *
+ * See file LICENSE for terms.
+ */
+#include <iostream>
+#include <mpi.h>
+#include <ucc/api/ucc.h>
+BEGIN_C_DECLS
+#include "core/ucc_mc.h"
+#include "utils/ucc_math.h"
+END_C_DECLS
+
+template<typename T>
+void init_buffer_host(void *buf, size_t count, int _value)
+{
+    T *ptr = (T *)buf;
+    T value = (T)_value;
+    for (int i = 0; i < count; i++) {
+        ptr[i] = value;
+    }
+}
+
+void init_buffer(void *buf, size_t count, ucc_datatype_t dt,
+                 ucc_memory_type_t mt, int value)
+{
+    if (mt != UCC_MEMORY_TYPE_HOST) {
+        std::cerr << "Unsupported mt\n";
+        MPI_Abort(MPI_COMM_WORLD, -1);
+    } else {
+        switch(dt) {
+        case UCC_DT_INT8:
+            init_buffer_host<int8_t>(buf, count, value);
+            break;
+        case UCC_DT_UINT8:
+            init_buffer_host<uint8_t>(buf, count, value);
+            break;
+        case UCC_DT_INT16:
+            init_buffer_host<int16_t>(buf, count, value);
+            break;
+        case UCC_DT_UINT16:
+            init_buffer_host<uint16_t>(buf, count, value);
+            break;
+        case UCC_DT_INT32:
+            init_buffer_host<int32_t>(buf, count, value);
+            break;
+        case UCC_DT_UINT32:
+            init_buffer_host<uint32_t>(buf, count, value);
+            break;
+        case UCC_DT_INT64:
+            init_buffer_host<int64_t>(buf, count, value);
+            break;
+        case UCC_DT_UINT64:
+            init_buffer_host<uint64_t>(buf, count, value);
+            break;
+        case UCC_DT_FLOAT32:
+            init_buffer_host<float>(buf, count, value);
+            break;
+        case UCC_DT_FLOAT64:
+            init_buffer_host<double>(buf, count, value);
+            break;
+        default:
+            std::cerr << "Unsupported dt\n";
+            MPI_Abort(MPI_COMM_WORLD, -1);
+            break;
+        }
+    }
+}
+
+template<typename T>
+ucc_status_t compare_buffers_fp(T *b1, T *b2, size_t count) {
+    for (int i = 0; i < count; i++) {
+        if (fabs((b1[i]-b2[i])/b1[i]) > 1e-5) {
```

Comment:
what if b1[i] is zero

Reply:
it is floating point value, it can't be exactly 0, it might be very small of course, but test buffers are initialized to avoid that. how else i can test PF buffers?

Reply:
ok, in worst case we will get nan or inf. There are a lot of ways how to do that, but maybe for our purposes this one is enough.

Reply:
if you know better way, plz share. quick googling didn't help me - i found essentially the same approach..

Reply:
One approach is to move denominator to right hand side, but it's slightly different inequality
see boost test description https://www.boost.org/doc/libs/1_32_0/libs/test/doc/components/test_tools/floating_point_comparison.html and corresponding implementation https://www.boost.org/doc/libs/1_32_0/boost/test/floating_point_comparison.hpp

Another suggestion is to compute max of b1 and b2 and use that max in denominator, that will allow to get consistent result regardless of b1 and b2 permutations
e.g. from Knuth book
```
bool approximatelyEqual(float a, float b, float epsilon)
{
    return fabs(a - b) <= ( (fabs(a) < fabs(b) ? fabs(b) : fabs(a)) * epsilon);
}

bool essentiallyEqual(float a, float b, float epsilon)
{
    return fabs(a - b) <= ( (fabs(a) > fabs(b) ? fabs(b) : fabs(a)) * epsilon);
}
```

And float test from googletest is a good reference as well

Reply:
👍 took approximatelyEqual